### PR TITLE
[XLA/tfcompile] Add Env::CreateUniqueFileName and use it in SaveGraph

### PR DESCRIPTION
**Pull request by rongjiecomputer**
Split part of `tensorflow::Env::LocalTempFilename` into `tensorflow::Env::CreateTempFilename` so that it can be used in `SaveGraph`.

This PR is to replace #15335.

I am terrible in creating descriptive function name, better name suggestion for `CreateTempFilename` is welcomed.

/cc @jlebar.

 #15213

### CODE REVIEW

File: tensorflow/core/platform/env.h [hunk 1]
```diff
@@ -218,6 +218,10 @@ class Env {
   /// Creates a local unique temporary file name. Returns true if success.
   bool LocalTempFilename(string* filename);
 
+  /// Create a unique file name starts with |prefix| and ends with |suffix|.
```
Comment (jlebar) [inline tensorflow/core/platform/env.h]: Nit: "Creates" (this is Google style, but also matches the surrounding code).

Comment (jlebar) [inline tensorflow/core/platform/env.h]: "that starts with"

File: tensorflow/core/platform/env.h [hunk 2]
```diff
@@ -218,6 +218,10 @@ class Env {
...
+  /// Returns true if success.
+  bool CreateTempFilename(string* prefix, const string& suffix);
```
Comment (jlebar) [inline tensorflow/core/platform/env.h]: How is this different from CreateLocalFilename?

Maybe what you actually want to do is add an optional suffix param to CreateLocalFilename?  Or even just call CreateLocalFilename inside of hlo_graph_dumper and just append the suffix to the returned filename?

File: tensorflow/core/platform/env.h (hunk 1 above)
Comment (rongjiecomputer) [reply tensorflow/core/platform/env.h]: Done

Comment (rongjiecomputer) [reply tensorflow/core/platform/env.h]: Done

File: tensorflow/core/platform/env.h (hunk 2 above)
Comment (rongjiecomputer) [reply tensorflow/core/platform/env.h]: > How is this different from CreateLocalFilename? (I mean, I can tell from reading the code, but not really from reading the header.)

Changed name to `CreateLocalFilename`.

> Maybe what you actually want to do is add an optional suffix param to CreateLocalFilename?

You mean `bool CreateTempFilename(string* prefix, const string& suffix = "")`?

> Or even just call CreateLocalFilename inside of hlo_graph_dumper and just append the suffix to the returned filename?

Then we will have to check if the new filename actually exists after appending suffix, repeating the work of `CreateLocalFilename`.

File: tensorflow/core/platform/env.h [hunk 3]
```diff
@@ -218,6 +218,10 @@ class Env {
   /// Creates a local unique temporary file name. Returns true if success.
   bool LocalTempFilename(string* filename);
 
+  /// Creates a local unique file name that starts with |prefix| and ends with
+  /// |suffix|. Returns true if success.
+  bool CreateLocalFilename(string* prefix, const string& suffix);
```
Comment (jlebar) [inline tensorflow/core/platform/env.h]: After pondering the implementation, I think perhaps CreateNewFileWithUniqueName might be a better name for this function.

File: tensorflow/core/platform/env.h [hunk 4]
```diff
@@ -279,7 +283,7 @@ class Env {
   // "version" should be the version of the library or NULL
   // returns the name that LoadLibrary() can use
   virtual string FormatLibraryFileName(const string& name,
-      const string& version) = 0;
+                                       const string& version) = 0;
```
Comment (jlebar) [inline tensorflow/core/platform/env.h]: Would prefer to leave these formatting changes to a different patch if possible.

File: tensorflow/core/platform/env.h [hunk 5]
```diff
@@ -346,6 +350,7 @@ class EnvWrapper : public Env {
                                const string& version) override {
     return target_->FormatLibraryFileName(name, version);
   }
+
```
Comment (jlebar) [inline tensorflow/core/platform/env.h]: Same.

**Review by jlebar — COMMENTED** [review]
Just a function name suggestion and then I think this will lgtm.

File: tensorflow/core/platform/env.h (hunk 2 above)
Comment (jlebar) [reply tensorflow/core/platform/env.h]: Sorry, I think I reviewed this before I saw this comment.

Anyway I like what you came up with.  :)

File: tensorflow/core/platform/env.h (hunk 3 above)
Comment (rongjiecomputer) [reply tensorflow/core/platform/env.h]: Done

File: tensorflow/core/platform/env.h (hunk 4 above)
Comment (rongjiecomputer) [reply tensorflow/core/platform/env.h]: Done. I blame clang-format for being too smart :)

File: tensorflow/core/platform/env_test.cc [hunk 6]
```diff
@@ -340,4 +341,18 @@ TEST_F(DefaultEnvTest, LocalTempFilename) {
...
 }
 
+TEST_F(DefaultEnvTest, CreateNewFileWithUniqueName) {
+  Env* env = Env::Default();
+
+  string prefix = "tempfile-prefix-";
+  string suffix = ".tmp";
+  string filename = prefix;
+
+  EXPECT_TRUE(env->CreateNewFileWithUniqueName(&filename, suffix));
```
Comment (jlebar) [inline tensorflow/core/platform/env_test.cc]: Since we're now creating the file, should we delete it too?  Especially since we're not putting it in /tmp.

(Sorry I missed this before.  This used to be fine when we were just returning a file*name*.)

File: tensorflow/core/platform/env.h (hunk 4 above)
Comment (jlebar) [reply tensorflow/core/platform/env.h]: FWIW I use the git-clang-format script to clang-format my changes without reformatting the whole file.  https://llvm.org/svn/llvm-project/cfe/trunk/tools/clang-format/git-clang-format

File: tensorflow/core/platform/env_test.cc (hunk 6 above)
Comment (rongjiecomputer) [reply tensorflow/core/platform/env_test.cc]: Both `CreateNewFileWithUniqueName` and `LocalTempFilename` do not create new file, they just create new file name, check if it exists in file system. If does not exist, return `true`, otherwise clear string and return `false`.

`CreateNewFileWithUniqueName` -> `CreateUniqueFileName`?

File: tensorflow/core/platform/env.h (hunk 4 above)
Comment (rongjiecomputer) [reply tensorflow/core/platform/env.h]: Thanks for the tips!

File: tensorflow/core/platform/env_test.cc (hunk 6 above)
Comment (jlebar) [reply tensorflow/core/platform/env_test.cc]: Ugh, sorry.  You're right, and that name sounds good.

Comment (rongjiecomputer) [reply tensorflow/core/platform/env_test.cc]: Done

**Review by jlebar — APPROVED** [review]
LGTM!  Thanks for being so patient with me.

**Comment by jlebar** [comment]
Jenkins, test this please.

(Am I doing this right?  I don't review too many OSS patches.)

**Comment by rongjiecomputer** [comment]
@jlebar I think you need to mention @tensorflow-jenkins 

See https://github.com/tensorflow/tensorflow/pull/15491#issuecomment-352861743

**Comment by jlebar** [comment]
Hm, maybe?  I see some people doing it without, e.g. https://github.com/tensorflow/tensorflow/pull/12503#issuecomment-330725991

Anyway, can't hurt.

@tensorflow-jenkins, test this please.

**Comment by rongjiecomputer** [comment]
@yifeif Can you help us to trigger CI tests?

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please
